File: Go/tic-tac-toe.go
package main

import (
	"fmt"
)

func main() {
	var board [3][3]int
	player1Turn := true
	for !checkWin(board, 1) && !checkWin(board, 2) {
		if checkDraw(board) {
			printBoard(board)
			fmt.Println("The game was a draw!")
			return
		}

		if player1Turn {
			fmt.Println("Please take a turn player 1")
		} else {
			fmt.Println("Please take a turn player 2")
		}

		fmt.Println("The board looks like this:")
		printBoard(board)
		fmt.Println("Type the position where you want to place your mark! [1-9]")

		index := 0
		for {
			_, err := fmt.Scanf("%d\n", &index) // read stdin and store the read number in index
			// validate input
			if err != nil || index < 1 || index > 9 {
				fmt.Println("Incorrect input. Please only type characters from 1 to 9")
			} else {
				break // exit loop if valid
			}
		}

		index -= 1 // turn 1-9 to 0-8 for use as indicies

		// if the field isn't already occupied
		if board[index/3][index%3] == 0 {
			// set field to correct state depending on who's turn it is
			if player1Turn {
				board[index/3][index%3] = 1
			} else {
				board[index/3][index%3] = 2
			}

			// other player's turn
			player1Turn = !player1Turn
		}
	}

	printBoard(board)
	if !player1Turn {
		fmt.Println("Player 1 won this game!")
	} else {
		fmt.Println("Player 2 won this game!")
	}
}

func checkWin(board [3][3]int, state int) bool {
	// check horizontal
	for i := 0; i < 3; i++ {
		if board[0][i] == state && board[1][i] == state && board[2][i] == state {
			return true
		}
	}

	// check vertical
	for i := 0; i < 3; i++ {
		if board[i][0] == state && board[i][1] == state && board[i][2] == state {
			return true
		}
	}

	// check diagonal
	if board[0][0] == state && board[1][1] == state && board[2][2] == state {
		return true
	}
	if board[2][0] == state && board[1][1] == state && board[0][2] == state {
		return true
	}

	return false
}

func checkDraw(board [3][3]int) bool {
	for i := 0; i < 3; i++ {
		for k := 0; k < 3; k++ {
			if board[i][k] == 0 {
				return false
			}
		}
	}
	return true
}

func printBoard(board [3][3]int) {
	for i := 0; i < 3; i++ {
		fmt.Println("+-+-+-+")
		for j := 0; j < 3; j++ {
			fmt.Printf("%c", '|')
			if board[i][j] == 0 {
				fmt.Printf("%c", ' ')
			} else if board[i][j] == 1 {
				fmt.Printf("%c", 'X')
			} else if board[i][j] == 2 {
				fmt.Printf("%c", 'O')
			}
		}
		fmt.Printf("%c\n", '|')
	}
	fmt.Println("+-+-+-+")
}

File: Rust/tic-tac-toe.rs
fn main() {
	let mut board = [0; 9];
	let mut player_1_turn = true;

	while !check_win(board) {
		if check_draw(board) {
			print_board(board);
			println!("This game was a draw!");
			return;
		}

		println!("Please take a turn player {}", (if player_1_turn { "1" } else { "2" }));
		println!("The board looks like this:");
		print_board(board);
		println!("Type the position where you want to place your mark! [1-9]");

		// parse user input
		let mut input: usize = loop {
			let mut buffer = String::new();
			match std::io::stdin().read_line(&mut buffer) {
				Ok(_) => {
					match buffer.trim().parse::<usize>() {
						Ok(x @ 1..=9) => break x,
						_ => println!("Incorrect input. Please only type characters from 1 to 9")
					}
				},
				Err(_) => println!("Incorrect input. Please only type characters from 1 to 9")
			}
		};
		

		input -= 1; // turn 1-9 to 0-8 for use as indicies

		// if the field isn't already occupied
		if board[input] == 0 {
			// set field to correct state depending on who's turn it is
			board[input] = if player_1_turn { 1 } else { 2 };

			// other player's turn
			player_1_turn = !player_1_turn;
		}
	}
	print_board(board);
	println!("Player {} won the game!", if !player_1_turn { "1" } else { "2" });
}

fn print_board(board: [u8; 9]) {
	for i in 0..3 {
		println!("+-+-+-+");
		for j in 0..3 {
			print!("|");
			match board[i * 3 + j] {
				0 => print!(" "),
				1 => print!("X"),
				2 => print!("O"),
				_ => print!("?"),
			}
		}
		println!("|");
	}
	println!("+-+-+-+");
}

fn check_win(board: [u8; 9]) -> bool {
	// check horizontal
	for i in 0..3 {
		// check X
		if board[0 + i * 3] == 1 && board[1 + i * 3] == 1 && board[2 + i * 3] == 1 {
			return true;
		}

		// check O
		if board[0 + i * 3] == 2 && board[1 + i * 3] == 2 && board[2 + i * 3] == 2 {
			return true;
		}
	}

	// check vertical
	for i in 0..3 {
		// check X
		if board[i] == 1 && board[i + 3] == 1 && board[i + 6] == 1 {
			return true;
		}

		// check O
		if board[i] == 2 && board[i + 3] == 2 && board[i + 6] == 2 {
			return true;
		}
	}

	// check diagonal X
	if (board[0] == 1 && board[4] == 1 && board[8] == 1) || board[2] == 1 && board[4] == 1 && board[6] == 1 {
		return true;
	}

	// check diagonal O
	if (board[0] == 2 && board[4] == 2 && board[8] == 2) || board[2] == 2 && board[4] == 2 && board[6] == 2 {
		return true;
	}

	return false;
}

fn check_draw(board: [u8; 9]) -> bool {
	for field in board {
		if field == 0 {
			return false;
		}
	}

	return true;
}